### DISCUSSION TITLE
How to replace fields in a segment?

### DISCUSSION THREAD

**Discussion by Coooz** [nextgenhealthcare/connect#4683] [Q&A]
I have this database in which I can store incoming HL7 messages. Each table represents a segment, each column represents a field plus component. Unfortunately this design does not allow storing repeating fields, so I came up with the following to provide for that. The segment is first scanned for repeating fields. This is for example in the incoming message (here I have put each field on a separate line to improve readability):

```
<PV1.9><PV1.9.1>02008</PV1.9.1><PV1.9.2>DAVIDS</PV1.9.2><PV1.9.3>SZILVIA</PV1.9.3></PV1.9>
<PV1.9><PV1.9.1>01116</PV1.9.1><PV1.9.2>SMITH</PV1.9.2><PV1.9.3>NURUDEEN</PV1.9.3><PV1.9.4>A</PV1.9.4></PV1.9>
<PV1.9><PV1.9.1>00976</PV1.9.1><PV1.9.2>JONES</PV1.9.2><PV1.9.3>DAVID</PV1.9.3><PV1.9.4>G</PV1.9.4></PV1.9>
```

My code then groups the respective values together per component, and to each value a sequence number is concatenated - with # as separator - that indicates the field index in which it occurred (again, the separate lines are done on purpose here to improve readability):

```
<PV1.9>
<PV1.9.1>02008#1#01116#2#00976#3</PV1.9.1>
<PV1.9.2>DAVIDS#1#SMITH#2#JONES#3</PV1.9.2>
<PV1.9.3>SZILVIA#1#NURUDEEN#2#DAVID#3</PV1.9.3>
<PV1.9.4>A#2#G#3</PV1.9.4>
</PV1.9>
```

This gives me each field and each component only once; so far, so good. But now I have to replace the original repeating fields in the segment with the modified fields (after this is done I can process the modified segment, and store the fields in the database). This code snippet should do so:

```
var regex = '<' + currentRepeatingField + '>.+<\/' + currentRepeatingField + '>'; //currentRepeatingField contains 'PV1.9'
var re = new RegExp(regex, 'g');
segment.replace(re, resultField); //resultField contains the modified PV1.9 field
```

re is the correct regular expression to select all <PV1.9>...</PV1.9> sections in the original segment from the first <PV1.9> to the last </PV1.9>, but even so the segment.replace-line gives an error: "TypeError: INVALID_CHARACTER_ERR: An invalid or illegal XML character is specified." I do not see what is wrong and which character is illegal. What am I doing wrong here?

Note:
Since it is apparently a "TypeError", it is quite possible that "segment" is not a string, and that might have something to do with it. After all, if I run this code

```
for each (segment in msg.children()) {
    logger.info('segment '+ segment.name().toString() + ': ' + segment.toString());
} 
```

I indeed have to specify segment _.toString()_ to have the segment show up in the log: merely _segment_ won't do it. But then: what does _segment_ itself look like, and how do I get original fields in this segment replaced with my modified fields?

Thanks!

**Comment by pacmano1** [comment]
That is somewhat of an atypical model.   Usually folks extract data as needed into normal tables columns (e.g. mrn goes into column mrn) and if a copy of the message is needed it is stored in its entirety in the database and retrieved and serialized as needed for future needs.  What is your use case to do it your way? I'm curious.

**Reply by Coooz** [reply]
No use case - I'm a Mirth Connect newbie and merely experimenting and learning by trying out.

**Comment by tonygermano** [comment]
In relation to @pacmano1 's question about your use case, and typically storing the raw hl7 message in its entirety, how does your database model handle subcomponents? Are you eventually reconstructing these hl7 messages or is this just to make the data easier to retrieve by other processes?

`segment` is most definitely not a string. It is an e4x xml object. If it were a string, you wouldn't be able to retrieve fields as properties.

I can't tell if this will work without seeing more of your code, but give this a try:
```js
segment.replace(currentRepeatingField, resultField)
```

This is assuming that `currentRepeatingField` is a string, and `resultField` is an xml object. If you have built `resultField` as a string it won't work correctly (but it won't error either.)

If you are new to e4x, here's a good place to start https://web.archive.org/web/20181120184304/https://wso2.com/project/mashup/0.2/docs/e4xquickstart.html

There are also many years worth of good information at the mirth forums, which is always worth a search https://forums.mirthproject.io/

**Reply by Coooz** [reply]
The database model does not handle subcomponents. Which is not (yet) a problem since I'm still only learning, and do so mainly by building and experimenting. The idea is indeed to make the data easier to retrieve by other processes.
`currentRepeatingField` and `resultField` are both strings.
Thanks for your information! I will certainly have a look at the sites you suggest.
